### PR TITLE
POC: yield loop - enabling yielding from training_step

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pl_examples/yield2.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        self.layer2 = torch.nn.Linear(32, 32)
+        self.head = torch.nn.Linear(32, 2)
+
+    # potential future directions
+    # 1) yield loss + optimizer
+    # 2) last statement must be a return
+    # 3) yield loss + extras for step_end and epoch_end
+    def training_step(self, batch, batch_idx, optimizer_idx=0):
+        loss0 = self.layer1(batch).sum()
+        yield loss0
```
Comment (carmocca) [inline pl_examples/yield2.py:37]: although this is cool, I don't expect many people to adopt this for 2 reasons:

- many python users are not familiar with yield
- harder to wrap your head around in general.
- people just love sprinkling if statements :joy:

(unrelated to your implementation)

Comment (justusschock) [reply pl_examples/yield2.py:37]: I kind of agree, this is also why we don't use contextmanagers in general if we can avoid it. Most people don't know how to deal with them. However, if we advertise it, I think a lot of people could adopt it, since it is way easier. And for the others, the "old" way would still be supported :)

Comment (awaelchli) [reply pl_examples/yield2.py:37]: totally agree, these points are well justified. 

at the very least this POC shows how one could possibly want to extend a lightning loop and shows us what barriers we have for enabling that in as few lines as possible.

Comment (tchaton) [reply pl_examples/yield2.py]: And this is a great opportunity to find `papercuts` within the Loops for customization too.
